### DISCUSSION TITLE
Bind a keymap to to perform multiple functions?

### DISCUSSION THREAD

**Discussion by voidstar-null** [YaLTeR/niri#897] [Q&A] [answered]
Is is possible to bind a single keymap to perform multiple functions in niri? 
For my usecase, I wish to bind the `Mod+Shift+Return`  to open a terminal window at workspace 1.
I tried everything I could think of, but did not succeed, So did I miss something from the wiki?

**Comment by YaLTeR** [comment, marked as answer]
Not at the moment. You can work around by binding something like `"sh" "-c" "niri msg action focus-workspace 1 && niri msg action spawn -- alacritty"`

**Reply by voidstar-null** [reply]
The problem with this is, with `workspace-auto-back-and-forth` enabled, launching alacritty would spawn it on the previously focused  workspace, if I am already on workspace 1.

**Comment by voidstar-null** [comment]
Is this feature planned ? cuz it currently seems like a workaround to achieve the described behaviour.

Also, I was trying to open a window as float using this keybind:
`Mod+Shift+Return repeat=false { spawn "sh" "-c" "niri msg action spawn -- foot && niri msg action toggle-window-floating"; }`

<strike>But the new float window does not gets focused.</strike> What am I doing wrong?

P.S : Happy New Year

**Reply by voidstar-null** [reply]
Apologies, this keybind doesn't even work as expected at all.

All it does is just spawn a terminall window and make the current window float T_T.

**Reply by YaLTeR** [reply]
This particular example won't work even with multiple bind actions, because it takes some time for the window to appear. Instead, spawn needs some kind of window rule argument that applies to the window. 

Alternatively, with foot specifically I think you can override the app id with a flag, then add a window rule to make that specific app id open-floating.

**Reply by voidstar-null** [reply]
> with foot specifically I think you can override the app id with a flag, then add a window rule to make that specific app id open-floating.

Yeah I also thought about this too. Thanks -)

> This particular example won't work even with multiple bind actions, because it takes some time for the window to appear. Instead, spawn needs some kind of window rule argument that applies to the window.

I wonder if passing some selected actions to the spawn command be doable in the possible future ? Seems like a nice and relevant feature.

**Reply by dyfrgi** [reply]
I think that Niri could track what is created by spawn via pid, but if the program called runs something other than via `exec` it would lose it. For instance, if you open a new foot window with `footclient`.

The stupidest thing that could possibly work would be to create a rule that applies to the next window created. You could make it more complex by having it time out, handling the case where the program fails to create a window at all.

A more complex alternative that also might work would be to set an env var and then check for that env var in the environment of processes that create new surfaces. It does have the problem of continuing to be passed along unless the process creating the surface knows about the env var, though, so if you open a terminal then things you launch from there will inherit it. If the env var contains a single-use token then future appearances can be ignored.

Both of these have a problem with "starting things not via exec", though. Suppose you trigger a launcher, which itself creates a window. How do you know that it shouldn't be consumed by the launcher? I suppose you could have *another* window rule that takes precedence over the token-consuming one.

I'd probably go with the stupid option, if I was trying to build this. Give it a total count and make it turn a window rule that's in the config file on and off. That way you can use other window rules to exclude stuff like launchers if you want to. For such windows, only the window opening properties would apply, since the rule would be turned off after a certain number of matches. Probably something like `match next-start-key="foo"` and then have `activate-start-key "foo" 2` to give it 2 activations.

Not sure how to add it to spawn, given how the argument parsing works. Might need to add parameter separation into it so that it can be `spawn <flags> <command>`. That or add the ability to have multiple commands within a single binding. Which is of course the original question.

It would be possible to make them "sticky" by having them set a property on the window when matched. That property could then be used within that same rule, or other rules. That's a small step up in complexity.
